github_app.go

```diff
@@ -0,0 +1,235 @@
+	}
+	authFields := s.getGithubAppAuth()
+	if authFields == nil {
+		// TODO EVG-19966: Return error here
+		grip.Debug(message.Fields{
+			"message": "no auth",
+			"owner":   owner,
+			"repo":    repo,
+			"ticket":  "EVG-19966",
+		})
+		return "", nil
+	}
+
+	retryConf := utility.NewDefaultHTTPRetryConf()
+	retryConf.MaxDelay = maxDelay
+	retryConf.BaseDelay = minDelay
+	retryConf.MaxRetries = maxRetries
+
+	httpClient := utility.GetHTTPRetryableClient(retryConf)
+	defer utility.PutHTTPClient(httpClient)
+
+	key, err := jwt.ParseRSAPrivateKeyFromPEM(authFields.privateKey)
+	if err != nil {
+		return "", errors.Wrap(err, "parsing private key")
+	}
+
+	itr := ghinstallation.NewAppsTransportFromPrivateKey(httpClient.Transport, authFields.AppId, key)
+	httpClient.Transport = itr
+	client := github.NewClient(httpClient)
+
+	installationID, err := getInstallationID(ctx, owner, repo)
+	if err != nil {
+		return "", errors.Wrapf(err, "getting cached installation id for '%s/%s'", owner, repo)
+	}
+	if installationID == 0 {
+		installationID, _, err := client.Apps.FindRepositoryInstallation(ctx, owner, repo)
+		if err != nil {
+			if strings.Contains(err.Error(), "404") {
+				return "", errors.Wrapf(gitHubAppNotInstalledError, "Installation id for '%s/%s' not found", owner, repo)
+			}
+			grip.Debug(message.WrapError(err, message.Fields{
+				"message": "error finding installation id",
+				"owner":   owner,
+				"repo":    repo,
+				"appId":   authFields.AppId,
+				"ticket":  "EVG-19966",
+			}))
+			return "", errors.Wrapf(err, "finding installation id for '%s/%s'", owner, repo)
+		}
+		if installationID == nil {
+			return "", errors.New(fmt.Sprintf("Installation id for '%s/%s' not found", owner, repo))
+		}
+
+		// Cache the installation ID for owner/repo.
+		hook := GitHubHook{
+			Owner:          owner,
+			Repo:           repo,
+			InstallationID: installationID.GetID(),
+		}
+		if err := hook.Upsert(ctx); err != nil {
+			return "", errors.Wrapf(err, "saving installation id for '%s/%s'", owner, repo)
+		}
+	}
+
+	token, _, err := client.Apps.CreateInstallationToken(ctx, installationID, opts)
+	if err != nil || token == nil {
+		return "", errors.Wrapf(err, "creating installation token for installation id: %d", installationID)
+	}
+	return token.GetToken(), nil
+}
+
+// HasGitHubApp returns true if the GitHub app is installed on given owner/repo.
+// Only returns an error if it is not a gitHubAppNotInstalledError.
+func (s *Settings) HasGitHubApp(ctx context.Context, owner, repo string, opts *github.InstallationTokenOptions) (bool, error) {
+	// Check cache for installation ID first.
+	installationID, err := getInstallationID(ctx, owner, repo)
+	if err != nil {
+		return false, errors.Wrapf(err, "getting cached installation id for '%s/%s'", owner, repo)
+	}
+	if installationID != 0 {
+		return true, nil
+	}
+
+	token, err := s.CreateInstallationToken(ctx, owner, repo, opts)
```

Comment:
We don't actually need the token here. we just need the installation ID. can we refactor this a bit? (or include it in the future refactor for thirdparty mentioned above)

Reply:
Can you add a TODO here: refactor to get installation ID instead of token